skglm/penalties/block_separable.py

```diff
@@ -153,3 +154,69 @@ def subdiff_distance(self, W, grad, ws):
     def is_penalized(self, n_features):
         """Return a binary mask with the penalized features."""
         return np.ones(n_features, bool_)
+
+
+spec_SparseGroupL1 = {
+    'alpha': float64,
+    'tau': float64,
+    'weights': float64[:]
+}
+
+
+@jitclass(spec_SparseGroupL1)
+class SparseGroupL1(BasePenalty):
+    r"""Sparse Group L1 penalty.
+
+    :math::
+
+        \alpha (\tau \Vert \beta \Vert_1 
+        + (1-\tau) \sum_g \omega_g \Vert \beta_g \Vert_2)
+
+    where :math:`\beta` is the vector of coefficients, :math:`\omega_g` are
+    the weights of the groups, :math:`\alpha` the intensity of the penalty, 
+    and :math:`\tau` the ratio individual/group penalty.
+
+    """
+
+    def __init__(self, alpha: float, tau: float,
+                 weights: np.ndarray, groups: np.ndarray):
+        self.alpha, self.tau = alpha, tau
+        self.weights, self.groups = weights, groups
+
+    def value(self, w: np.ndarray) -> float:
+        alpha, tau = self.alpha, self.tau
+        weights, groups = self.weights, self.groups
+
+        pen_l1_term = alpha * tau * norm(w, ord=1)
+
+        pen_group_term = 0.
+        for g in groups:
+            if weights[g] == np.inf:
+                continue
+            pen_group_term += alpha * (1-tau) * weights[g] * norm(w[g], ord=2)
+
+        return pen_l1_term + pen_group_term
+
+    def prox_1feat(self, value: np.ndarray, stepsize: float, j: int) -> np.ndarray:
+        tau, weight_j = self.tau, self.weights[j]
+        mask_zero_coefs = value != 0
+
+        # ST for each features
+        value[mask_zero_coefs] = ST_vec(
```

Comment:
This is pretty dangerous here as I do an in-place assignment and hence the changes are reflected directly into the original object. I didn't want to create a new instance as looking back at the code, I find out that we constantly feed ``prox_1feat`` a new vector (``value`` is obtained through an arithmetic operation between two arrays)

Reply:
The changes are reflected directly in the original instance.

Let's say we want to return an array of the iterates $\beta^{(k)}$, then following this design, the elements will be altered.

Reply:
you should use copy when storing all iterates.
value is the result after a gradient step so it's not an iterate
the current design is more memory efficient